Add jit_stack_size option

alubbe opened the pull request:
This is a quick suggestion on how a jit_stack_size option could be implemented. I lack the actual C skills to pull this off, and I would first like to discuss the direction of this approach.

Allocating and freeing jit_stacks for each regex is too costly according to http://www.pcre.org/original/doc/html/pcrejit.html#SEC9, and it seems safe to reuse the same stack if the code runs single-threaded according to that same FAQ.

My idea would thus be to have a "global" jit_stack, initially set to 32K (the current default), and to have a lua-land function akin to this

``` lua
function ngx.re.opt(option, value)
  if option == "jit_stack_size" then
    ngx_http_lua_ffi_set_jit_stack_size(value)
  end
end
```

``` c
void 
ngx_http_lua_ffi_set_jit_stack_size(int size)
{
  pcre_jit_stack_free(jit_stack);
  int min_size = MIN(32 * 1024, size);
  jit_stack = pcre_jit_stack_alloc(min_size, size);
}
```

Every time a regex is compiled in ngx_http_lua_regex.c, we just add

``` c
pcre_assign_jit_stack(sd, NULL, jit_stack);
```

What remains to be discussed is where `jit_stack` lives. It could be a mutable C global, it could be part of the Lua state or we could instantiate it on every compile, which would be more expensive, but assuming that most users combine "j" with "o", it might not make a dent after all.

Feedback is most welcome.

agentzh commented:
@alubbe The direction is correct, I think.

The global `jit_stack` should live in the `ngx_http_lua_main_conf_t` struct. The C globals are not safe enough for HUP reload.

alubbe commented:
Updated this PR and added https://github.com/openresty/lua-resty-core/pull/44. 
Where should `jit_stack` be instantiated intially? In the C code or from lua?

agentzh commented:
@alubbe It should be initialized in the C function `ngx_http_lua_create_main_conf`, I guess?

alubbe commented:
Added. I feel like that 32 * 1024 magic number should live somewhere, something like MIN_JIT_STACK_SIZE. Where would its place be?

Also, aside from adding tests, together with https://github.com/openresty/lua-resty-core/pull/44 is there anything missing?

alubbe commented on src/ngx_http_lua_module.c:
```diff
@@ -798,6 +798,7 @@ ngx_http_lua_create_main_conf(ngx_conf_t *cf)
 #if (NGX_PCRE)
     lmcf->regex_cache_max_entries = NGX_CONF_UNSET;
     lmcf->regex_match_limit = NGX_CONF_UNSET;
+    lmcf->jit_stack = pcre_jit_stack_alloc(32*1024, 32*1024);
```
We should also check that this is not NULL. What's the idiomatic way of throwing errors here?

agentzh replied:
@alubbe Just return NULL here when the allocation fails.

agentzh replied:
@alubbe Also, we should guard this piece with macros when PCRE is built without JIT or an older version of PCRE is used.

alubbe commented on src/ngx_http_lua_regex.c:
```diff
@@ -2478,6 +2480,22 @@ ngx_http_lua_ffi_max_regex_cache_size(void)
+    lmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle,
+                                               ngx_http_lua_module);
+    if (lmcf == NULL) {
+        return;
```
Similarly, how should this error be handled?

agentzh replied:
`lmcf` cannot be `NULL`. So this check is unneeded.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -2478,6 +2480,22 @@ ngx_http_lua_ffi_max_regex_cache_size(void)
     }
     return (uint32_t) lmcf->regex_cache_max_entries;
 }
+
```
Style: two blank lines are used to separate C function definitions.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -2478,6 +2480,22 @@ ngx_http_lua_ffi_max_regex_cache_size(void)
+    int                          min_size;
+    ngx_http_lua_main_conf_t    *lmcf;
+
+    min_size = MIN(32 * 1024, size);
```
We need a named constant to replace such magic numbers.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -2478,6 +2480,22 @@ ngx_http_lua_ffi_max_regex_cache_size(void)
+        return;
+    }
+    pcre_jit_stack_free(lmcf->jit_stack);
+    lmcf->jit_stack = pcre_jit_stack_alloc(min_size, size);
```
Error checks needed here.

alubbe replied:
From my understanding of the docs there is no return code to check for errors http://pcre.org/original/doc/html/pcre_jit_stack_alloc.html

agentzh replied:
@alubbe Allocation can always fail. You should check if the pointer is NULL.

agentzh commented:
@alubbe So it won't work if lua-resty-core is not used?

agentzh commented:
@alubbe Your PR is currently failing Travis CI checks BTW. It does not even compile.

alubbe commented:
I fixed the build, added some guards and extracted the magic number.
Is there a way to share the `#define` parts between the files, in particular `LUA_HAVE_PCRE_JIT` and `NGX_LUA_RE_MIN_JIT_STACK_SIZE`?

alubbe commented:
Tbh, this PR is as much a discussion as it is an implementation attempt. This feature is probably simple enough to be brought into the non-resty-core world via the lua5.1-c api, but I have exactly zero experience with that.

alubbe commented:
For tests, we could use something like the below (thanks to https://github.com/JuliaLang/julia/issues/8278)
```lua
ngx.re.opt("jit_stack_size", 128 * 1024) -- the match below will fail without this
local subject = [[71.163.72.113 - - [30/Jul/2014:16:40:55 -0700] "GET example.com/thevacantwall/wp-content/uploads/2013/02/DSC_006421.jpg HTTP/1.1" 200 492513 "http://images.search.yahoo.com/images/view;_ylt=AwrB8py9gdlTGEwADcSjzbkF;_ylu=X3oDMTI2cGZrZTA5BHNlYwNmcC1leHAEc2xrA2V4cARvaWQDNTA3NTRiMzYzY2E5OTEwNjBiMjc2YWJhMjkxMTEzY2MEZ3BvcwM0BGl0A2Jpbmc-?back=http%3A%2F%2Fus.yhs4.search.yahoo.com%2Fyhs%2Fsearch%3Fei%3DUTF-8%26p%3Dapartheid%2Bwall%2Bin%2Bpalestine%26type%3Dgrvydef%26param1%3D1%26param2%3Dsid%253Db01676f9c26355f014f8a9db87545d61%2526b%253DChrome%2526ip%253D71.163.72.113%2526p%253Dgroovorio%2526x%253DAC811262A746D3CD%2526dt%253DS940%2526f%253D7%2526a%253Dgrv_tuto1_14_30%26hsimp%3Dyhs-fullyhosted_003%26hspart%3Dironsource&w=588&h=387&imgurl=occupiedpalestine.files.wordpress.com%2F2012%2F08%2F5-peeking-through-the-wall.jpg%3Fw%3D588%26h%3D387&rurl=http%3A%2F%2Fwww.stopdebezetting.com%2Fwereldpers%2Fcompare-the-berlin-wall-vs-israel-s-apartheid-wall-in-palestine.html&size=49.0KB&name=...+%3Cb%3EApartheid+wall+in+Palestine%3C%2Fb%3E...+%7C+Or+you+go+peeking+through+the+%3Cb%3Ewall%3C%2Fb%3E&p=apartheid+wall+in+palestine&oid=50754b363ca991060b276aba291113cc&fr2=&fr=&tt=...+%3Cb%3EApartheid+wall+in+Palestine%3C%2Fb%3E...+%7C+Or+you+go+peeking+through+the+%3Cb%3Ewall%3C%2Fb%3E&b=0&ni=21&no=4&ts=&tab=organic&sigr=13evdtqdq&sigb=19k7nsjvb&sigi=12o2la1db&sigt=12lia2m0j&sign=12lia2m0j&.crumb=.yUtKgFI6DE&hsimp=yhs-fullyhosted_003&hspart=ironsource" "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1985.125 Safari/537.36]]
local regex = [[([\d\.]+) ([\w.-]+) ([\w.-]+) (\[.+\]) "([^"\r\n]*|[^"\r\n\[]*\[.+\][^"]+|[^"\r\n]+.[^"]+)" (\d{3}) (\d+|-) ("(?:[^"]|\")+)"? ("(?:[^"]|\")+)"?]]
local matches, err = match(subject, regex, "j")
return ngx.say(matches and matches[0] or err)
```
As of right now, it segfaults, but my C knowledge is so extremely limited that I don't even know how retrieve a meaningful error message. But I feel that we are very close.

alubbe commented:
I'd love to finish this PR, but I'll need a little bit of help on why it segfaults and how to best write the test that I put together.

alubbe commented:
Okay so one large part of the reason it segfaulted was that I was not using the new jit_stack in the ffi functions, but the larger problem is https://github.com/openresty/lua-nginx-module/pull/782/files#diff-5a62a11d0a2e8e55e4c5a7feaac0edd1R2502
This call to free the previous stack fails - I cannot figure out why. If I comment it out (and accept leaking 32kb of memory, assuming you only call the new helper once during worker init or similar) everything works great and the test that would be failing (https://github.com/openresty/lua-nginx-module/pull/782#issuecomment-223569981) passes. You can also reduce the jit_stack_size down to 32kb again and make the test case fail again.

Hence, we need to figure out why `pcre_jit_stack_free(lmcf->jit_stack);` fails, wrap my tests into the official suite and maybe add a non-jit helper function to set a new jit_stack. I definitely need help with this ;)

alubbe commented:
Through some trial and error, we now have a working ffi and non-ffi implementation as well as new and passing tests - with one big caveat: this line segfaults everything https://github.com/openresty/lua-nginx-module/pull/782/files#diff-5a62a11d0a2e8e55e4c5a7feaac0edd1R1940

If you could point me in the direction of how to figure this one out, I can remove that final roadblock to finishing up the PR.

alubbe commented:
Just a thought: depending on what you prefer, we could also merge the two test cases into one, thereby reusing the subject and regex variables.

agentzh commented on src/ngx_http_lua_module.c:
```diff
@@ -28,6 +28,18 @@
+#   define LUA_HAVE_PCRE_JIT 0
+#endif
+
+#endif
```
These macros are already defined in `common.h`, so why the duplication here?

alubbe replied:
Because I have absolutely no idea how C works ;) Will remove them!

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1923,11 +1927,51 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+
+    lmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle,
+                                               ngx_http_lua_module);
+    // pcre_jit_stack_free(lmcf->jit_stack); // Y U NO WORK?
```
Style: no C++-style line comments please.

agentzh replied:
Here is a memory leak. The problem you were having should be that some cached `pcre_extra` objects still keep references to the old JIT stack.

The right solution might be to keep a reference count in the JIT stack and free the stack in places like `ngx_http_lua_regex_free_study_data` when the reference count of the JIT stack reaches zero. It deserves special care though otherwise there will be memory leaks.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1923,11 +1927,51 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+    option = luaL_checklstring(L, 1, NULL);
+    value = luaL_checkint(L, 2);
+
+    if (strcmp(option, "jit_stack_size") == 0) {
```
Should use the nginx variant instead of calling `strcmp` directly.

agentzh commented on t/120-re-find.t:
```diff
@@ -891,3 +891,63 @@ not matched!
+            -- regex is taken from https://github.com/JuliaLang/julia/issues/8278
+            ngx.re.opt("jit_stack_size", 128 * 1024)
+            local s = [[71.163.72.113 - - [30/Jul/2014:16:40:55 -0700] "GET emptymind.org/thevacantwall/wp-content/uploads/2013/02/DSC_006421.jpg HTTP/1.1" 200 492513 "http://images.search.yahoo.com/images/view;_ylt=AwrB8py9gdlTGEwADcSjzbkF;_ylu=X3oDMTI2cGZrZTA5BHNlYwNmcC1leHAEc2xrA2V4cARvaWQDNTA3NTRiMzYzY2E5OTEwNjBiMjc2YWJhMjkxMTEzY2MEZ3BvcwM0BGl0A2Jpbmc-?back=http%3A%2F%2Fus.yhs4.search.yahoo.c ...
+            local regex = [[([\d\.]+) ([\w.-]+) ([\w.-]+) (\[.+\]) "([^"\r\n]*|[^"\r\n\[]*\[.+\][^"]+|[^"\r\n]+.[^"]+)" (\d{3}) (\d+|-) ("(?:[^"]|\")+)"? ("(?:[^"]|\")+)"?]]
```
Style: shall we split these long lines via `..`?

alubbe replied:
I'm torn on that issue. I feel that splitting long lines helps when those lines are meant to be understood and reviewed. Here, I feel as if those two variables might as well be called `very_long_string` and `very_complicated_regex`. When someone reads this test, I'd like them to focus on the title of the test, the 5-6 lines of code and the expected output. For more details about the string and regex, the reader could be referred to the github issue in the link above.

agentzh commented on src/ngx_http_lua_common.h:
```diff
@@ -22,6 +22,19 @@
+#   define LUA_HAVE_PCRE_JIT 0
+#endif
+
+#endif
```
Now that we define these macros in this `common.h` header that is included in all the other compilation units (directly or indirectly), we should remove the duplicate definitions from `regex.c` I think?

alubbe replied:
Done.

agentzh commented:
@alubbe I prefer separate smaller test cases which are easier to debug. I don't really care about duplication in the tests :)

Thanks for your hard work on this PR. Will you look into my latest comments? Thanks!

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1923,11 +1927,51 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+    if (strcmp(option, "jit_stack_size") == 0) {
+        ngx_http_lua_set_jit_stack_size(value);
+    }
+
```
Need to throw out an error for unrecognized option names instead of silently ignoring them. Also need a test case for this.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -363,6 +365,7 @@ ngx_http_lua_ngx_re_match_helper(lua_State *L, int wantcaps)
         old_pool = ngx_http_lua_pcre_malloc_init(pool);
 
         sd = pcre_study(re_comp.regex, PCRE_STUDY_JIT_COMPILE, &msg);
+        pcre_assign_jit_stack(sd, NULL, lmcf->jit_stack);
```
Maybe we can avoid this C function call when the user has never set any jit stack size explicitly? We could make `lmcf->jit_stack` default to `NULL` and always tests the nullity of `lmcf->jit_stack` before calling `pcre_assign_jit_stack`. Thoughts?

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1923,11 +1927,51 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+                                               ngx_http_lua_module);
+    // pcre_jit_stack_free(lmcf->jit_stack); // Y U NO WORK?
+    lmcf->jit_stack = pcre_jit_stack_alloc(NGX_LUA_RE_MIN_JIT_STACK_SIZE,
+                                           size);
```
We should handle the case that `pcre_jit_stack_alloc` returns a `NULL` when running out of memory.

alubbe commented:
Defaulting `lmcf->jit_stack` to `NULL` was a very good idea - the allocation was not working this early anyways. The new approach allows freeing the jit_stack just fine, so the memory leak is now fixed. I've also added more and better error messages.

The issue remains how to handle changing the jit_stack_size when there are already compiled regex with `"jo"` depending on the old jit_stack. My vote would be to document that this is not supported, i.e., that `ngx.re.opt("jit_stack_size", ...)` MUST be called before any regex are called with `"jo"`. I see the need to set it once during the initialization, but I do not see a use case where you would want to change it on the fly.

What do you think?

agentzh commented:
@alubbe I'm fine with disallowing changing jit stack size after any regexes are compiled and cached. But I think documentation is not enough. We should throw out a Lua exception when the user tries to do this instead of leading to mysterious nginx process crashes.

agentzh commented:
@alubbe BTW, I'm glad that ngx_lua gets your feet wet in the C world :)

alubbe commented:
It's an exciting trip in C land, but my rust experience is definitely helping.
Just wanted to let you know I might not have time this and next week, hopefully I can finish the PR thereafter. Also, I have deployed both PRs onto our production system, where ngx.re.opt is run in init_by_lua and a lot of regex are used in every transaction. So far, it has been without a single error or segfault, having already processed several gigabytes of data.

agentzh commented:
@alubbe Good to know. Thanks!

alubbe commented:
Rebased and updated this PR - feedback most welcome!

alubbe commented:
From my point of view, this PR is finished. What are your thoughts?

alubbe commented:
ping @agentzh 
if you're busy just let me know

alubbe commented:
with the release out of the door, what are the chances of this being merged for the next one @agentzh ?

alubbe commented:
@agentzh I'd really appreciate an update from you on this PR. It's been running in production without hiccups for months now and I'd love to see it merged rather than living as a separate patch.

alubbe commented:
1.11.2.2 being just around the corner, could you give an update on how likely it is that this PR will be reviewed and merged? I'd be happy to put more work into it if there are concrete issues to tackle.

agentzh commented:
@alubbe Sorry for the delay on my side. Massively distracted by other things.

@doujiang24 Will you have a look at this PR please? Many thanks!

doujiang24 commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,76 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+
+        if (result == -1) {
+            return luaL_error(L, "Changing jit stack size is not allowed when "
+                                "regexs have already been compiled and cached");
```
style: should align up with `L`

doujiang24 replied:
style: better add a blank line after this line

doujiang24 commented on t/120-re-find.t:
```diff
@@ -891,3 +891,103 @@ not matched!
+=== TEST 31: default jit_stack_size too small
+--- config
+    location /re {
+        content_by_lua '
```
We usually use content_by_lua_block for new test case :)

doujiang24 commented on t/120-re-find.t:
```diff
@@ -891,3 +891,103 @@ not matched!
+content_by_lua(nginx.conf:43):2: unrecognized option name for ngx.re.opt
+--- no_error_log
+[error]
+
```
None or one blank line is needed here :)

doujiang24 commented:
@alubbe Nice job!
@agentzh LGTM:)

alubbe commented:
Updated.
If this PR is fine, I've also opened a PR for resty.core https://github.com/openresty/lua-resty-core/pull/44
Both this and the resty.core PR have run fine on several billion requests in the last months, but I am getting errors on the test suite that I need some help with: https://github.com/openresty/lua-resty-core/pull/44#issuecomment-234528584

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+                                 "some regexs have already been compiled and "
+                                 "cached");
+
+        } else if (result == -2) {
```
We sholud use `if` instead of `else if` here since the previous branch never falls through.

agentzh commented on src/ngx_http_lua_regex.h:
```diff
@@ -13,7 +13,8 @@
 
 #if (NGX_PCRE)
-void ngx_http_lua_inject_regex_api(lua_State *L);
+void      ngx_http_lua_inject_regex_api(lua_State *L);
```
Style: we should use a single space after `void` here.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+        return luaL_error(L, "expecting two arguments, but got %d", nargs);
+    }
+
+    option = luaL_checklstring(L, 1, NULL);
```
We should use fetch the string length here.

alubbe replied:
Sorry, what do you mean?

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+    option = luaL_checklstring(L, 1, NULL);
+    value = luaL_checkint(L, 2);
+
+    if (ngx_strcmp(option, "jit_stack_size") == 0) {
```
We should use `ngx_strncmp` here to avoid testing `\0` upon every character of `option` here.

agentzh replied:
We need to handle the case when JIT is disabled in the PCRE build.

alubbe replied:
Which length do you want to use here? If we use the length of `option`, `"j"` will match, and if we use the length of `"jit_stack_size"`, then `"jit_stack_sizefoobar"` will match.
I expect this function to be called once during init_worker, so we might not need an optimization here.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+                                               ngx_http_lua_module);
+
+    if (lmcf->regex_cache_entries > 0) {
+        return -1;
```
Better avoid using magical numbers like `-1` and `-2` here. We could reuse existing constants like `NGX_ERROR`, `NGX_DECLINED` and `NGX_BUSY` here, if applicable.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+    lmcf->jit_stack = pcre_jit_stack_alloc(NGX_LUA_RE_MIN_JIT_STACK_SIZE,
+                                           size);
+
+    if (!lmcf->jit_stack) {
```
Style: for NULL pointers test, we usually write this instead:

```C
if (lmcf->jit_stack == NULL) {
```

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+    }
+
+    if (lmcf->jit_stack) {
+        pcre_jit_stack_free(lmcf->jit_stack);
```
It seems that we only free the stack upon setting a new JIT stack size. How about the last allocated stack? Logically we should also free that last stack. Maybe we should always register a pool cleanup handler for the memory pool that allocates `lmcf` here?

alubbe replied:
I don't understand nginx's internals well enough to follow. My intention is simple - register a stack once globally (per worker) and reuse it for every call. Then, if you change the jit_stack_size, free the old stack and set a new one as the global one. Is that not what the current code does?

agentzh replied:
@alubbe One thing in my mind is that we will make `ngx.re` work in `init_by_lua*`, in which case there might be a memory leak in the nginx master process upon HUP reload.

alubbe replied:
We use it in `init_by_lua` and have not seen any noticeable memory leaks across HUP reloads. That's not to say that we aren't leaking but we have not noticed it. Anyways, how can we solve it? I genuinely do not know enough about nginx's memory pools and layout, esp. across reloads.

agentzh replied:
@alubbe Have you use valgrind to check leaks?

alubbe replied:
Here is what happens around a HUP reload:
```
==45767== Memcheck, a memory error detector
==45767== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==45767== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==45767== 
==45767== Conditional jump or move depends on uninitialised value(s)
==45767==    at 0x526BE39: str_fastcmp (lj_str.c:58)
==45767==    by 0x526BE39: lj_str_new (lj_str.c:166)
==45767==    by 0x527F15A: lua_pushlstring (lj_api.c:578)
==45767==    by 0x52E10EB: emptybuffer (lib_aux.c:177)
==45767==    by 0x52E10EB: luaL_pushresult (lib_aux.c:221)
==45767==    by 0x52E12D5: luaL_gsub (lib_aux.c:163)
==45767==    by 0x50656F: ngx_http_lua_set_path.isra.7.constprop.19 (ngx_http_lua_util.c:147)
==45767==    by 0x50CED8: ngx_http_lua_new_state (ngx_http_lua_util.c:271)
==45767==    by 0x50CED8: ngx_http_lua_init_vm (ngx_http_lua_util.c:3721)
==45767==    by 0x50115A: ngx_http_lua_init (ngx_http_lua_module.c:731)
==45767==    by 0x485BC8: ngx_http_block (ngx_http.c:312)
==45767==    by 0x465188: ngx_conf_handler (ngx_conf_file.c:427)
==45767==    by 0x466272: ngx_conf_parse (ngx_conf_file.c:283)
==45767==    by 0x463374: ngx_init_cycle (ngx_cycle.c:274)
==45767==    by 0x44EC9F: main (nginx.c:276)
==45767== 
==45767== Warning: set address range perms: large range [0x1607c000, 0x3547c000) (defined)
==45792== 
==45789== 
==45788== 
==45789== HEAP SUMMARY:
==45789==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45789==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45789== 
==45792== HEAP SUMMARY:
==45792==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45792==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45792== 
==45788== HEAP SUMMARY:
==45788==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45788==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45788== 
==45790== 
==45790== HEAP SUMMARY:
==45790==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45790==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45790== 
==45791== 
==45791== HEAP SUMMARY:
==45791==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45791==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45791== 
==45793== 
==45793== HEAP SUMMARY:
==45793==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45793==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45793== 
==45786== 
==45786== HEAP SUMMARY:
==45786==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45786==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45786== 
==45787== 
==45787== HEAP SUMMARY:
==45787==     in use at exit: 4,374,347 bytes in 3,780 blocks
==45787==   total heap usage: 17,370 allocs, 13,590 frees, 68,947,724 bytes allocated
==45787== 
==45794== 
==45794== HEAP SUMMARY:
==45794==     in use at exit: 351,435 bytes in 3,780 blocks
==45794==   total heap usage: 17,374 allocs, 13,594 frees, 64,925,148 bytes allocated
==45794== 
==45795== 
==45795== HEAP SUMMARY:
==45795==     in use at exit: 351,435 bytes in 3,780 blocks
==45795==   total heap usage: 17,370 allocs, 13,590 frees, 64,924,812 bytes allocated
==45795== 
==45790== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45790==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45790==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45790==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45790==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45790==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45790==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45790==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45790==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45790==    by 0x44EEB1: main (nginx.c:367)
==45790== 
==45790== LEAK SUMMARY:
==45790==    definitely lost: 960,000 bytes in 1 blocks
==45790==    indirectly lost: 3,280,000 bytes in 2 blocks
==45790==      possibly lost: 0 bytes in 0 blocks
==45790==    still reachable: 134,347 bytes in 3,777 blocks
==45790==         suppressed: 0 bytes in 0 blocks
==45790== Reachable blocks (those to which a pointer was found) are not shown.
==45790== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45790== 
==45790== For counts of detected and suppressed errors, rerun with: -v
==45790== Use --track-origins=yes to see where uninitialised values come from
==45790== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45786== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45786==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45786==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45786==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45786==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45786==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45786==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45786==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45786==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45786==    by 0x44EEB1: main (nginx.c:367)
==45786== 
==45786== LEAK SUMMARY:
==45786==    definitely lost: 960,000 bytes in 1 blocks
==45786==    indirectly lost: 3,280,000 bytes in 2 blocks
==45786==      possibly lost: 0 bytes in 0 blocks
==45786==    still reachable: 134,347 bytes in 3,777 blocks
==45786==         suppressed: 0 bytes in 0 blocks
==45786== Reachable blocks (those to which a pointer was found) are not shown.
==45786== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45786== 
==45786== For counts of detected and suppressed errors, rerun with: -v
==45786== Use --track-origins=yes to see where uninitialised values come from
==45786== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45787== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45787==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45787==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45787==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45787==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45787==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45787==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45787==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45787==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45787==    by 0x44EEB1: main (nginx.c:367)
==45787== 
==45787== LEAK SUMMARY:
==45787==    definitely lost: 960,000 bytes in 1 blocks
==45787==    indirectly lost: 3,280,000 bytes in 2 blocks
==45787==      possibly lost: 0 bytes in 0 blocks
==45787==    still reachable: 134,347 bytes in 3,777 blocks
==45787==         suppressed: 0 bytes in 0 blocks
==45787== Reachable blocks (those to which a pointer was found) are not shown.
==45787== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45787== 
==45787== For counts of detected and suppressed errors, rerun with: -v
==45787== Use --track-origins=yes to see where uninitialised values come from
==45787== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45793== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45793==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45793==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45793==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45793==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45793==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45793==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45793==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45793==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45793==    by 0x44EEB1: main (nginx.c:367)
==45793== 
==45793== LEAK SUMMARY:
==45793==    definitely lost: 960,000 bytes in 1 blocks
==45793==    indirectly lost: 3,280,000 bytes in 2 blocks
==45793==      possibly lost: 0 bytes in 0 blocks
==45793==    still reachable: 134,347 bytes in 3,777 blocks
==45793==         suppressed: 0 bytes in 0 blocks
==45793== Reachable blocks (those to which a pointer was found) are not shown.
==45793== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45793== 
==45793== For counts of detected and suppressed errors, rerun with: -v
==45793== Use --track-origins=yes to see where uninitialised values come from
==45793== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45792== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45792==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45792==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45792==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45792==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45792==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45792==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45792==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45792==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45792==    by 0x44EEB1: main (nginx.c:367)
==45792== 
==45792== LEAK SUMMARY:
==45792==    definitely lost: 960,000 bytes in 1 blocks
==45792==    indirectly lost: 3,280,000 bytes in 2 blocks
==45792==      possibly lost: 0 bytes in 0 blocks
==45792==    still reachable: 134,347 bytes in 3,777 blocks
==45792==         suppressed: 0 bytes in 0 blocks
==45792== Reachable blocks (those to which a pointer was found) are not shown.
==45792== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45792== 
==45792== For counts of detected and suppressed errors, rerun with: -v
==45792== Use --track-origins=yes to see where uninitialised values come from
==45792== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45788== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45788==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45788==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45788==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45788==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45788==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45788==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45788==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45788==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45788==    by 0x44EEB1: main (nginx.c:367)
==45788== 
==45788== LEAK SUMMARY:
==45788==    definitely lost: 960,000 bytes in 1 blocks
==45788==    indirectly lost: 3,280,000 bytes in 2 blocks
==45788==      possibly lost: 0 bytes in 0 blocks
==45788==    still reachable: 134,347 bytes in 3,777 blocks
==45788==         suppressed: 0 bytes in 0 blocks
==45788== Reachable blocks (those to which a pointer was found) are not shown.
==45788== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45788== 
==45788== For counts of detected and suppressed errors, rerun with: -v
==45788== Use --track-origins=yes to see where uninitialised values come from
==45788== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45791== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45791==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45791==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45791==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45791==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45791==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45791==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45791==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45791==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45791==    by 0x44EEB1: main (nginx.c:367)
==45791== 
==45791== LEAK SUMMARY:
==45791==    definitely lost: 960,000 bytes in 1 blocks
==45791==    indirectly lost: 3,280,000 bytes in 2 blocks
==45791==      possibly lost: 0 bytes in 0 blocks
==45791==    still reachable: 134,347 bytes in 3,777 blocks
==45791==         suppressed: 0 bytes in 0 blocks
==45791== Reachable blocks (those to which a pointer was found) are not shown.
==45791== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45791== 
==45791== For counts of detected and suppressed errors, rerun with: -v
==45791== Use --track-origins=yes to see where uninitialised values come from
==45791== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45789== 4,240,000 (960,000 direct, 3,280,000 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45789==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45789==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45789==    by 0x470148: ngx_event_process_init (ngx_event.c:692)
==45789==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45789==    by 0x47864A: ngx_worker_process_cycle (ngx_process_cycle.c:734)
==45789==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45789==    by 0x47917F: ngx_start_worker_processes (ngx_process_cycle.c:358)
==45789==    by 0x479715: ngx_master_process_cycle (ngx_process_cycle.c:130)
==45789==    by 0x44EEB1: main (nginx.c:367)
==45789== 
==45789== LEAK SUMMARY:
==45789==    definitely lost: 960,000 bytes in 1 blocks
==45789==    indirectly lost: 3,280,000 bytes in 2 blocks
==45789==      possibly lost: 0 bytes in 0 blocks
==45789==    still reachable: 134,347 bytes in 3,777 blocks
==45789==         suppressed: 0 bytes in 0 blocks
==45789== Reachable blocks (those to which a pointer was found) are not shown.
==45789== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45789== 
==45789== For counts of detected and suppressed errors, rerun with: -v
==45789== Use --track-origins=yes to see where uninitialised values come from
==45789== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45794== 217,088 (118,784 direct, 98,304 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45794==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45794==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45794==    by 0x47011D: ngx_event_process_init (ngx_event.c:685)
==45794==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45794==    by 0x478B02: ngx_cache_manager_process_cycle (ngx_process_cycle.c:1114)
==45794==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45794==    by 0x479333: ngx_start_cache_manager_processes (ngx_process_cycle.c:396)
==45794==    by 0x47971F: ngx_master_process_cycle (ngx_process_cycle.c:132)
==45794==    by 0x44EEB1: main (nginx.c:367)
==45794== 
==45794== LEAK SUMMARY:
==45794==    definitely lost: 118,784 bytes in 1 blocks
==45794==    indirectly lost: 98,304 bytes in 2 blocks
==45794==      possibly lost: 0 bytes in 0 blocks
==45794==    still reachable: 134,347 bytes in 3,777 blocks
==45794==         suppressed: 0 bytes in 0 blocks
==45794== Reachable blocks (those to which a pointer was found) are not shown.
==45794== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45794== 
==45794== For counts of detected and suppressed errors, rerun with: -v
==45794== Use --track-origins=yes to see where uninitialised values come from
==45794== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
==45795== 217,088 (118,784 direct, 98,304 indirect) bytes in 1 blocks are definitely lost in loss record 803 of 803
==45795==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==45795==    by 0x47478E: ngx_alloc (ngx_alloc.c:22)
==45795==    by 0x47011D: ngx_event_process_init (ngx_event.c:685)
==45795==    by 0x4780BD: ngx_worker_process_init (ngx_process_cycle.c:900)
==45795==    by 0x478B02: ngx_cache_manager_process_cycle (ngx_process_cycle.c:1114)
==45795==    by 0x476F78: ngx_spawn_process (ngx_process.c:198)
==45795==    by 0x47944A: ngx_start_cache_manager_processes (ngx_process_cycle.c:413)
==45795==    by 0x47971F: ngx_master_process_cycle (ngx_process_cycle.c:132)
==45795==    by 0x44EEB1: main (nginx.c:367)
==45795== 
==45795== LEAK SUMMARY:
==45795==    definitely lost: 118,784 bytes in 1 blocks
==45795==    indirectly lost: 98,304 bytes in 2 blocks
==45795==      possibly lost: 0 bytes in 0 blocks
==45795==    still reachable: 134,347 bytes in 3,777 blocks
==45795==         suppressed: 0 bytes in 0 blocks
==45795== Reachable blocks (those to which a pointer was found) are not shown.
==45795== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==45795== 
==45795== For counts of detected and suppressed errors, rerun with: -v
==45795== Use --track-origins=yes to see where uninitialised values come from
==45795== ERROR SUMMARY: 64 errors from 2 contexts (suppressed: 0 from 0)
```

agentzh replied:
@alubbe Just FYI, it's much easier to just take advantage of the valgrind test mode of the Test::Nginx test  scaffold as documented below:

https://openresty.gitbooks.io/programming-openresty/content/testing/test-modes.html#_valgrind_mode

agentzh replied:
@alubbe Thanks for your verification! I'll double check on my side :)

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+    const char   *option;
+    int           nargs;
+    int           value = 0;
+    int           result;
```
We usually use the C variable name `rc` instead of `result` here.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+{
+    const char   *option;
+    int           nargs;
+    int           value = 0;
```
This initialization is unneeded since `value` is always assigned before being used here.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+        return 0;
+    }
+
+    return luaL_error(L, "unrecognized option name for ngx.re.opt");
```
Better use shorter error message like "unrecognized option name". The context can always be obtained from the backtrace of the Lua error anyway.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
 
 
+ngx_int_t
+ngx_http_lua_set_jit_stack_size(int size)
```
This C function should be guarded by `LUA_HAVE_PCRE_JIT`. Better test with PCRE builds without JIT enabled to be sure here.

agentzh commented:
@alubbe I think we also need some documentation for this new feature ;)

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,78 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+
+        if (result == -1) {
+            return luaL_error(L, "Changing jit stack size is not allowed when "
+                                 "some regexs have already been compiled and "
```
Style: the continued line should align up with the first argument on the previous line.

alubbe commented:
I'll add some documentation later. This builds fine without pcre jit on my machine.

agentzh commented:
@alubbe It's better to avoid squashing commits during review. So it's easier for us to see what your new changes are instead of always scanning the whole changeset everytime you push :)

agentzh commented:
@alubbe Please do not amend your existing commits in your branch, which makes tracking your new changes much harder. Thank you. Before merging your changes, I'll always squash your commits in your PR anyway.

agentzh commented:
@alubbe Please always commit your new changes in new commits and push them into the same branch.

alubbe commented:
Sorry about that, I did not realise you were looking at these commits as they were coming in. I had some issues with travis so I rolled back and re-commited to trigger new test runs, not to make it harder on you :)

agentzh commented:
@alubbe BTW, I think you can just restart finished Travis CI jobs in the web UI of Travis CI site without making artificial changes to the git repo. I've always been doing that myself ;)

agentzh commented on README.markdown:
```diff
@@ -6084,6 +6085,40 @@ This feature was first introduced in the `v0.2.1rc15` release.
 
 [Back to TOC](#nginx-api-for-lua)
 
+ngx.re.opt
```
To quote the comment at the beginning of this file:

```
<!---
Don't edit this file manually! Instead you should generate it by using:
    wiki2markdown.pl doc/HttpLuaModule.wiki
-->
```

In short, you are editing the wrong file :)

agentzh commented on t/120-re-find.t:
```diff
@@ -891,3 +891,101 @@ not matched!
+--- response_body
+content_by_lua(nginx.conf:43):2: unrecognized option name
+--- no_error_log
+[error]
```
I think we need a test case for success calls of `ngx.re.opt("jit_stack_size")` in a request handler, like `content_by_lua*`.

agentzh replied:
Also need a test case for some regexes have indeed been compiled but *never* cached when `ngx.re.opt("jit_stack_size")` is called.

alubbe replied:
I've updated TEST 33 to incorporate both points.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,12 +1924,89 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+        return NGX_ERROR;
+    }
+
+    return 0;
```
Better use `NGX_OK` here instead of `0`.

alubbe commented:
Hmm so the new test now fails the test suite, but works fine if you run an isolated `nginx.conf`. Are there any cached regex run in an initializer somewhere that is not visible from the test file?

agentzh commented:
@alubbe It seems that there is bigger issue with `TEST 33: jit_stack_size change disallowed once regex cache is populated` in `t/120-re-find.t`. Valgrind reports the following error:

```
t/120-re-find.t .. # I found ONLY: maybe you're debugging?
TEST 33: jit_stack_size change disallowed once regex cache is populated
t/120-re-find.t .. 1/8 ==120331== Invalid read of size 8
==120331==    at 0x4032757: ???
==120331==    by 0x8FAFD6F: ???
==120331==    by 0x8FAFD6F: ???
==120331==    by 0x8FAFD7A: ???
==120331==    by 0x8FC422F: ???
==120331==  Address 0x8fb6ff8 is 8 bytes inside a block of size 32 free'd
==120331==    at 0x4C29CF0: free (vg_replace_malloc.c:530)
==120331==    by 0x41E860: ngx_destroy_pool (ngx_palloc.c:79)
==120331==    by 0x4597FA: ngx_http_free_request (ngx_http_request.c:3516)
==120331==    by 0x4598AF: ngx_http_close_request (ngx_http_request.c:3427)
==120331==    by 0x45ACC6: ngx_http_finalize_connection (ngx_http_request.c:2577)
==120331==    by 0x45B248: ngx_http_finalize_request (ngx_http_request.c:2459)
==120331==    by 0x455978: ngx_http_core_content_phase (ngx_http_core_module.c:1379)
==120331==    by 0x450633: ngx_http_core_run_phases (ngx_http_core_module.c:856)
==120331==    by 0x450741: ngx_http_handler (ngx_http_core_module.c:839)
==120331==    by 0x45BDA8: ngx_http_process_request (ngx_http_request.c:1916)
==120331==    by 0x45C184: ngx_http_process_request_headers (ngx_http_request.c:1348)
==120331==    by 0x45C499: ngx_http_process_request_line (ngx_http_request.c:1027)
==120331==    by 0x45C8D3: ngx_http_wait_request_handler (ngx_http_request.c:504)
==120331==    by 0x444B68: ngx_epoll_process_events (ngx_epoll_module.c:900)
==120331==    by 0x43B2D9: ngx_process_events_and_timers (ngx_event.c:242)
==120331==    by 0x443BBE: ngx_single_process_cycle (ngx_process_cycle.c:309)
==120331==    by 0x41D200: main (nginx.c:364)
==120331==  Block was alloc'd at
==120331==    at 0x4C28BF6: malloc (vg_replace_malloc.c:299)
==120331==    by 0x43EB63: ngx_alloc (ngx_alloc.c:22)
==120331==    by 0x41E6AF: ngx_malloc (ngx_palloc.c:140)
==120331==    by 0x41E8F1: ngx_palloc (ngx_palloc.c:123)
==120331==    by 0x4DB96F: ngx_http_lua_pcre_malloc (ngx_http_lua_pcrefix.c:35)
==120331==    by 0x5C5EA63: sljit_allocate_stack (sljitUtils.c:239)
==120331==    by 0x5C5EA63: pcre_jit_stack_alloc (pcre_jit_compile.c:11607)
==120331==    by 0x4CE67E: ngx_http_lua_set_jit_stack_size (ngx_http_lua_regex.c:1949)
==120331==    by 0x4CE719: ngx_http_lua_ngx_re_opt (ngx_http_lua_regex.c:1982)
==120331==    by 0x5465177: lj_BC_FUNCC (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==120331==    by 0x4D5E61: ngx_http_lua_run_thread (ngx_http_lua_util.c:1005)
==120331==    by 0x4D99BD: ngx_http_lua_content_by_chunk (ngx_http_lua_contentby.c:120)
==120331==    by 0x4D9B92: ngx_http_lua_content_handler_inline (ngx_http_lua_contentby.c:310)
==120331==    by 0x4D957C: ngx_http_lua_content_handler (ngx_http_lua_contentby.c:222)
==120331==    by 0x45596D: ngx_http_core_content_phase (ngx_http_core_module.c:1379)
==120331==    by 0x450633: ngx_http_core_run_phases (ngx_http_core_module.c:856)
==120331==    by 0x450741: ngx_http_handler (ngx_http_core_module.c:839)
==120331==    by 0x45BDA8: ngx_http_process_request (ngx_http_request.c:1916)
==120331==    by 0x45C184: ngx_http_process_request_headers (ngx_http_request.c:1348)
==120331==    by 0x45C499: ngx_http_process_request_line (ngx_http_request.c:1027)
==120331==    by 0x45C8D3: ngx_http_wait_request_handler (ngx_http_request.c:504)
==120331==    by 0x444B68: ngx_epoll_process_events (ngx_epoll_module.c:900)
==120331==    by 0x43B2D9: ngx_process_events_and_timers (ngx_event.c:242)
==120331==    by 0x443BBE: ngx_single_process_cycle (ngx_process_cycle.c:309)
==120331==    by 0x41D200: main (nginx.c:364)
==120331==
```

agentzh commented:
@alubbe It seems that you need to call `ngx_http_lua_pcre_malloc_init` and `ngx_http_lua_pcre_malloc_done` around your `pcre_jit_stack_alloc` and `pcre_jit_stack_free` calls. This is because nginx hijacks `pcre_malloc` and `pcre_free` so we need to be super careful about which nginx memory pool we are currently allocating through the current PCRE API calls.

alubbe commented:
Thanks for clarifying. I've updated the PR and the test 33 now seems to run, somewhat. Similar to the failures in the resty.core code, there are some regex somewhere that get compiled and cached before our test code runs, even though, in isolation, all tests pass successfully. Any idea of what's going on?

agentzh commented:
@alubbe Are you aware of `repeated_each`? Seems like it is just because of that.

agentzh commented:
@alubbe I've thought more about it. Maybe it's a better idea to just add this `opt` API function to the `ngx.re` module of `lua-resty-core`? See

https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/re.md#readme

This way, we can avoid two parallel versions of the API implementation.

alubbe commented:
So should we just disable repeated_each for Test 33?

Could you elaborate a little bit on when features should be put into the `ngx.re` module as opposed to directly into `ngx.re`? Or is the plan to build/release new features for resty-core/FFI-based only? All resty-core features are two parallel versions of the API implementation, aren't they?

agentzh commented:
@alubbe 

> So should we just disable repeated_each for Test 33?

Not really, the `grep_error_log_out` section supports different expected outputs for each individual repeated requests, just do not use pcall to catch the Lua error so that we can test the error logs. You can find similar examples in the existing test suite.

> Could you elaborate a little bit on when features should be put into the ngx.re module
> as opposed to directly into ngx.re?

Eventually, all the `ngx.re.*` API functions will be moved into the `ngx.re` module. So it's not a good
idea to add new API functions to the `ngx.re` global table since it adds new backward-compatibility burdens.

> Or is the plan to build/release new features for resty-core/FFI-based only?

Yes. We're moving away from the standard Lua 5.1 interpreter compatbility. It's too much a burden nowadays when FFI is favored.

> All resty-core features are two parallel versions of the API implementation, aren't they?

Well, the goal is to have only one implementation and make existing `ngx.re.*` API functions just use our new FFI-based implementation under the hood.

alubbe commented:
Hmm, I cannot seem to get `grep_error_log_out` to work - help?

Regarding `ngx.re`, I agree to push it into `resty.core` only. I'll work on that as soon as we get this test passing.

alubbe commented:
I am in need of some help to get the test working.
Once it does, I will migrate things over to resty.core. This PR will then just add some C code, but no Lua interface to call it, since the resty.core PR would do that - correct?

agentzh commented:
@alubbe I had a quick glance at your failed tests. You might consider rerunning those failed tests in the valgrind test mode to see if there's anything memory issues. Also, please see other uses of `grep_error_log_out` in the existing test suite for examples.

alubbe commented:
@agentzh I've spent another few hours trying to understand the issue. I'm afraid my command of C debugging tools is not sufficient. Like we already know, the code itself works, but the test is not doing what I'd like it to. Also, studying the other tests using `grep_error_log_out` have not been able to help me out. You can see from my last commits my desparate attempts to make it work, but to no avail.
It looks as if I'm in need of some hand holding here.

Could you help me in writing a proper test? I'd really appreciate that.

In return, once the test passes, I'll separate everything so that this PR contains only the C code and the resty.core PR contains the lua interface and documentation.

agentzh commented:
@alubbe Thank you for your hard work!

@doujiang24 Will you please help @alubbe with his failing tests? ;)

doujiang24 commented:
@alubbe  How about this patch?
Cause `luaL_error` will raise an error and broke the request.

```patch
@@ -960,11 +960,17 @@ to: 1563
 --- config
     location /re {
         content_by_lua_block {
+            local function set_opt()
+                return ngx.re.opt("jit_stack_size", 128 * 1024)
+            end
+            local ok, err = pcall(set_opt)
+            if not ok then
+                ngx.log(ngx.INFO, err)
+            end
             local s = "hello, 1234"
             local from, to = ngx.re.find(s, "(hello world)|([0-9])", "jo")
             ngx.say("from: ", from)
             ngx.say("to: ", to)
-            ngx.re.opt("jit_stack_size", 128 * 1024)
         }
     }
 --- request
@@ -974,10 +980,12 @@ from: 8
 to: 8

 --- grep_error_log eval
-Changing jit stack size is not allowed when some regexs have already been compiled and cached
+qr/Changing jit stack size is not allowed when some regexs have already been compiled and cached/

 --- grep_error_log_out eval
 ["", "Changing jit stack size is not allowed when some regexs have already been compiled and cached\n"]
```

alubbe commented:
Had to rebase because of a new test case in 120-re-find that was causing merge conflicts

@doujiang24 I've applied your idea in my latest commit, let's see what travis says :)

alubbe commented:
Finally - all tests pass 👍 
When removing the non-ffi lua API, should this PR contain any tests at all or are the ones that will be in the resty.core PR sufficient?

agentzh commented:
@alubbe Leaving tests in lua-resty-core should be sufficient.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -2210,6 +2256,10 @@ ngx_http_lua_ffi_compile_regex(const unsigned char *pat, size_t pat_len,
                                                ngx_http_lua_module);
 
+    if (sd && lmcf->jit_stack) {
+        pcre_assign_jit_stack(sd, NULL, lmcf->jit_stack);
```
I think this should be protected by the macro `LUA_HAVE_PCRE_JIT`.

agentzh commented on src/ngx_http_lua_regex.h:
```diff
@@ -14,6 +14,7 @@
 
 #if (NGX_PCRE)
 void ngx_http_lua_inject_regex_api(lua_State *L);
+ngx_int_t ngx_http_lua_ffi_set_jit_stack_size(int size);
```
Ditto.

agentzh replied:
Because this C API is part of the ABI, I think we should always make this C API function present but always return an error in that C function when PCRE JIT is not present.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -826,6 +824,10 @@ ngx_http_lua_ngx_re_gmatch(lua_State *L)
         sd = pcre_study(re_comp.regex, PCRE_STUDY_JIT_COMPILE, &msg);
 
+        if (lmcf->jit_stack) {
+            pcre_assign_jit_stack(sd, NULL, lmcf->jit_stack);
```
At this point, `sd` might be `NULL` we should be prepared for that here.

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -364,6 +358,10 @@ ngx_http_lua_ngx_re_match_helper(lua_State *L, int wantcaps)
         sd = pcre_study(re_comp.regex, PCRE_STUDY_JIT_COMPILE, &msg);
 
+        if (lmcf->jit_stack) {
+            pcre_assign_jit_stack(sd, NULL, lmcf->jit_stack);
```
At this point, `sd` might be `NULL` we should be prepared for that here.

alubbe commented:
The travis failure is due to an unrelated flaky test. I do not have the rights to restart the build, could someone trigger that?

agentzh commented:
@alubbe Well, we cannot upgrade PCRE since some of our tests are relying on the specific behavior of PCRE 8.33.

alubbe commented:
No worries, I reverted that commit and now everything passes.

alubbe commented:
Regarding pcre 8.33, is there a way to upgrade the tests? I would feel more comfortable if the test suite ran the same versions we are using in production.

I had a quick glance at the tests that rely on 8.33 - they seem to be the ones trying to hit the match limit (error: pcre_exec() failed: -8) but instead they now work. How did pcre change that those tests no longer "fail" as they should? Is there a way to upgrade the tests?

agentzh commented:
@alubbe It just needs manual tuning for the right internal maxmatch threshold of PCRE, which changes due to new PCRE internal updates (like optimizations). You are welcome to update our tests for the latest PCRE release as long as they are still testing what they are supposed to :)

alubbe commented:
I'll take a look! Are you at all familiar with the pcre codebase and do you have any pointers as to where the which code bits / which functions have changed?

agentzh commented:
@alubbe No, there are a *lot* of changes since PCRE 8.33. I think you have to do your own experiments with those match limit magic numbers.

alubbe commented:
Updated - https://github.com/openresty/lua-nginx-module/pull/950

doujiang24 commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,6 +1924,55 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+
+#else  /* LUA_HAVE_PCRE_JIT */
+
+    return NGX_ERROR;
```
We'd better use another error code here, like `NGX_DECLINED`?
So that we can handle this case better in the Lua code :)

doujiang24 commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,6 +1924,55 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+}
+
+
+
```
two blank lines needed here, but found three :)

alubbe commented:
I've updated both PRs to handle missing pcre jit support.

alubbe commented:
Are there any remaining open points?

alubbe commented:
@agentzh What's the status here, is there anything left for me to do?

agentzh commented:
@alubbe Sorry for the delay on my side. Will look into this soon. Thanks!

alubbe commented:
@agentzh Any chance of this making the next release?

agentzh commented on src/ngx_http_lua_regex.c:
```diff
@@ -1922,6 +1924,54 @@ ngx_http_lua_ngx_re_sub_helper(lua_State *L, unsigned global)
+
+#else  /* LUA_HAVE_PCRE_JIT */
+
+    return NGX_DECLINED;
```
Returning `NGX_ERROR` here is better?

agentzh replied:
Also, it's even better to produce an nginx error log message here.

agentzh commented:
@alubbe Unfortunately the merge window of 1.11.2.3 is already closed. Sorry about that. Need to kick out a new release ASAP.

alubbe commented:
Ah too bad. I'll make sure to bug you once it's out so that it might get merged sooner rather than later ;)

alubbe commented:
Test failure seems to be an unrelated flaky test. Could someone trigger a re-run?

alubbe commented:
Perfect, both PRs are rebased and green. Any remaining open points @agentzh ?

agentzh commented:
@alubbe I think it's mostly fine now. Will get to this as soon as I can manage. Sorry, have more urgent things to take care of atm.

agentzh commented:
@alubbe I'm terribly sorry for the long delay on my side! Merged with the following extra patch:

```diff
diff --git a/src/ngx_http_lua_module.c b/src/ngx_http_lua_module.c
index 72f934d0..7b17ac41 100644
--- a/src/ngx_http_lua_module.c
+++ b/src/ngx_http_lua_module.c
@@ -814,6 +814,7 @@ ngx_http_lua_create_main_conf(ngx_conf_t *cf)
      *      lmcf->running_timers = 0;
      *      lmcf->watcher = NULL;
      *      lmcf->regex_cache_entries = 0;
+     *      lmcf->jit_stack = NULL;
      *      lmcf->shm_zones = NULL;
      *      lmcf->init_handler = NULL;
      *      lmcf->init_src = { 0, NULL };
diff --git a/src/ngx_http_lua_regex.c b/src/ngx_http_lua_regex.c
index 20a45cdc..80519ecd 100644
--- a/src/ngx_http_lua_regex.c
+++ b/src/ngx_http_lua_regex.c
@@ -1926,7 +1926,7 @@ error:

 ngx_int_t
 ngx_http_lua_ffi_set_jit_stack_size(int size, u_char *errstr,
-                                    size_t *errstr_size)
+    size_t *errstr_size)
 {
 #if LUA_HAVE_PCRE_JIT

diff --git a/src/ngx_http_lua_regex.h b/src/ngx_http_lua_regex.h
index 9752827d..03dffb80 100644
--- a/src/ngx_http_lua_regex.h
+++ b/src/ngx_http_lua_regex.h
@@ -15,7 +15,7 @@
 #if (NGX_PCRE)
 void ngx_http_lua_inject_regex_api(lua_State *L);
 ngx_int_t ngx_http_lua_ffi_set_jit_stack_size(int size, u_char *errstr,
-                                              size_t *errstr_size);
+    size_t *errstr_size);
 #endif


```

alubbe commented:
Perfect, it's been a long journey :D
Thanks again for all of your input and patience with me!
